[H] Move the splitting of licenses to an isolate

Hixie opened the pull request:
This improves (from horrific to terrible) the performance of the
license screen. It also introduces a feature in the foundation layer
to make using isolates for one-off computations easier.

The real problem that remains with this, though, is that transfering
data between isolates is a stop-the-world operation and can take an
absurd amount of time (far more than a few milliseconds), so we still
skip frames.

More work thus remains to be done.

Hixie committed:
Move the splitting of licenses to an isolate

This improves (from horrific to terrible) the performance of the
license screen. It also introduces a feature in the foundation layer
to make using isolates for one-off computations easier.

The real problem that remains with this, though, is that transfering
data between isolates is a stop-the-world operation and can take an
absurd amount of time (far more than a few milliseconds), so we still
skip frames.

More work thus remains to be done.

Hixie commented:
Actually those numbers are for debug builds. For profile builds it's still in the ~320ms range.

yjbanov commented on packages/flutter/lib/src/foundation/isolates.dart:
```diff
@@ -0,0 +1,66 @@
+/// Instances of [ComputeCallback] must be top-level functions, not closures or
+/// instance or static methods of classes.
+///
+/// {@macro flutter.foundation.compute.limitations}
```
I'm not familiar with the `{@` syntax, but this line looks a little suspicious. I'd expect either `{@endmacro}` to close the macro started above (like `{@endtemplate` below), or some content to follow this line 😕

Hixie replied:
As far as I can tell, `{@template name} ... {@endtemplate}` declares a macro named "name", and `{@macro name}` inserts the macro named "name".

yjbanov commented on packages/flutter/lib/src/foundation/isolates.dart:
```diff
@@ -0,0 +1,66 @@
+/// from isolates. These limitations constrain the values of `Q` and `R` that
+/// are possible. See the discussion at [SendPort.send].
+/// {@endtemplate}
+Future<R> compute<Q, R>(ComputeCallback<Q, R> callback, Q message) async {
```
I'm wondering if returning `Stream<R>` would be more conducive to designing more effective solutions to jank. If each element of the stream is sent as a separate message, it would be processed in its own event cycle thus not blocking the UI (e.g. it wouldn't jank an animating progress indicator). For example, we could have `_parseLicenses` return `Stream<LicenseEntry>` and deserialize one entry per event instead of parsing the whole list of entries in one event. Another benefit is that `Stream` can be cancelled, causing the isolate to shut down (e.g. to implement a timeout). We shouldn't lose in generality as `Future<R>` can be emulated using `Stream<R>` (e.g. `stream.first`).

Hixie replied:
Yeah I tried adding Stream support just before submitting this (and going to bed), but in the few minutes I had before bed time, I couldn't get it to really work well. I can push at it some more.

yjbanov replied:
Actually, we can add support for Stream and Future later, since they are not supported right now. We could dynamically check the type of the returned value, and if we find that it's a Stream or a Future, we send them down specialized code paths.

Hixie replied:
ok, sounds good

yjbanov commented on packages/flutter/lib/src/foundation/isolates.dart:
```diff
@@ -0,0 +1,66 @@
+/// instance or static methods of classes.
+///
+/// {@macro flutter.foundation.compute.limitations}
+typedef R ComputeCallback<Q, R>(Q message);
```
It would probably be very useful to support `Future` and especially `Stream` as return values (see discussion below about streams).

Hixie replied:
Good point about Future. I'll see how I can make that work.

yjbanov commented on packages/flutter/lib/src/foundation/isolates.dart:
```diff
@@ -0,0 +1,66 @@
+/// {@macro flutter.foundation.compute.types}
+///
+/// Instances of [ComputeCallback] must be top-level functions, not closures or
+/// instance or static methods of classes.
```
Why not static methods? AFAICT, isolates support them.

Hixie replied:
Oh, cool. I may have misread the docs.

yjbanov reviewed (commented):
Enjoying it indeed! 😃

yjbanov commented:
![lgtm](https://user-images.githubusercontent.com/211513/32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d.png)

Hixie commented:
will land on green then keep trying to make this better.

Hixie committed:
- Add profile instrumentation to the isolate compute() method
- Add profile instrumentation to the LicensePage
- Add profile instrumentation to the scheduleTask method
- Make scheduleTask support returning a value
- Make the license page builder logic use scheduled tasks so that it doesn't blow the frame budget

Hixie commented:
Actually, new plan. I added a bunch of instrumentation so I could figure out what was going on better, and that led to another obvious optimisation, and that led to me restructuring our task system.

Long story short, this patch got even more interesting, PTAL? :-)

The new logic is in a separate commit for easier review.

sroddy commented:
Hi @Hixie,  I've just taken a look at your work on this PR. It gave me a lot of insights on how to deal with some performance issues I'm having inside our app. However there is something that is not clear to me: I saw that you scheduled the tasks with a Priority.animation flag. Does this mean that if there is any kind of animation inside a view (maybe an continuous infinite animation), any other task that doesn't have Priority.animation is locked forever waiting for it to stop, even if there is room in the frame budget to potentially execute other tasks?

yjbanov commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -418,6 +430,7 @@ class _LicensePageState extends State<LicensePage> {
     setState(() {
       _loaded = true;
     });
+    Timeline.timeSync('Build scheduled', () { }, flow: Flow.end(flow.id));
```
Hmm, it seems what we really want is something like https://api.dartlang.org/stable/1.24.3/dart-html/Performance/mark.html, but strangely it's only available in `dart:html`. Have you filed a feature request?

Hixie replied:
Filed https://github.com/dart-lang/sdk/issues/31950

yjbanov commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -33,14 +33,20 @@ set timeDilation(double value) {
+/// Signature for frame-related callbacks from the scheduler.
 ///
 /// The timeStamp is the number of milliseconds since the beginning of the
 /// scheduler's epoch. Use timeStamp to determine how far to advance animation
```
nit: while you're here could you wrap "timeStamp" in backticks?

Hixie replied:
good catch. done.

yjbanov commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -33,14 +33,20 @@ set timeDilation(double value) {
 
+/// Signature for [Scheduler.scheduleTask] callbacks.
+///
+/// The type argument is the task's return value. Consider [void] if the task
```
nit: I'd say "The type argument `T` is the..."

Hixie replied:
done

yjbanov commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -253,23 +274,45 @@ abstract class SchedulerBinding extends BindingBase with ServicesBinding {
+  final PriorityQueue<_TaskEntry<dynamic>> _taskQueue = new HeapPriorityQueue<_TaskEntry<dynamic>>(_taskSorter);
 
-  /// Schedules the given `task` with the given `priority`.
+  /// Schedules the given `task` with the given `priority` and returns the
```
nit: s/"the a"/"da"/ 😉

Hixie replied:
fixed

yjbanov commented:
@sroddy, yeah I might be useful to add a few controls for better scheduling, such as a timeout (after which a task is executed eagerly, as if `Priority.touch`) and an estimated task duration, so if we finished with a frame early enough to fit the task we'd run it during animation without risking missing frames. That said, it should probably be a separate PR.

Hixie commented:
@sroddy that's right. Generally speaking, Priority.animation and priorities derived from it are the right priorities to use. Priority.touch is the "high priority" one and you generally don't want that one.

Hixie commented:
I don't understand why the sample code analyzer finds an error here; it doesn't find it on my local machine...

sroddy commented:
@yjbanov currently it’s also important to make sure that in any part of the framework or other first and third-party libraries tasks are scheduled with a priority lower than “animation” otherwise there is a concrete risk that that code wouldn’t be executed if an app uses animations intensively (I’m thinking about games, but that’s not the only example I have in mind).

Maybe it could be worth to have another priority that is lower than animation (so that animation task are anyways executed first) but tasks with that priority are not blocked by a lightweight animations.

In any case I have the feeling that without a timeout the way this method of the scheduler works is a bit dangerous and could lead to potentially hard to debug bugs (especially because the default priority is not animation and people would need to really understand what’s going on before figuring out what’s blocking something)

Hixie commented:
There's about 100,000 priority levels that are below "touch" but still above "animation". You can add ints to Priority.animation to make whatever groups of priorities you like.

For example, `Priority.animation + 100` is going to execute before `Priority.animation + 10`.

sroddy commented:
Thanks for the clarification. There’s definitely enough levels to play with :-)

I think however that it would be worth to add the timeout and weight parameters in order to avoid potential “deadlocks” and support a better management of the frame budget

Hixie commented:
Please file bugs for those feature requests (ideally describing the use cases). Not sure when we'll get to them but they're definitely things we want to track. (Also happy to review PRs of course.)
